spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java

```diff
@@ -767,4 +824,270 @@ public void close() throws IOException {
+              .build();
+
+      SparkFileWriterFactory writerFactoryWithoutRow =
+          SparkFileWriterFactory.builderFor(table)
+              .dataFileFormat(format)
+              .dataSchema(writeSchema)
+              .dataSparkType(dsSchema)
+              .deleteFileFormat(format)
+              .positionDeleteSparkType(deleteFileType)
+              .build();
+
+      return new DeleteWriter(
+          table,
+          writerFactoryWithRow,
+          writerFactoryWithoutRow,
+          deleteFileFactory,
+          targetFileSize,
+          dsSchema);
+    }
+  }
+
+  private static class DeleteWriter implements DataWriter<InternalRow> {
+    private final ClusteredPositionDeleteWriter<InternalRow> writerWithRow;
+    private final ClusteredPositionDeleteWriter<InternalRow> writerWithoutRow;
+    private final PositionDelete<InternalRow> positionDelete;
+    private final FileIO io;
+    private final Map<Integer, PartitionSpec> specs;
+    private final InternalRowWrapper partitionRowWrapper;
+    private final Map<Integer, StructProjection> partitionProjections;
+    private final int specIdOrdinal;
+    private final Option<Integer> partitionOrdinalOption;
+    private final int fileOrdinal;
+    private final int positionOrdinal;
+    private final int rowOrdinal;
+    private final int rowSize;
+
+    private boolean closed = false;
+
+    /**
+     * Writer for position deletes metadata table.
+     *
+     * <p>Delete files need to either have 'row' as required field, or omit 'row' altogether, for
+     * delete file stats accuracy Hence, this is a fanout writer, redirecting rows with null 'row'
+     * to one delegate, and non-null 'row' to another
+     *
+     * @param table position deletes metadata table
+     * @param writerFactoryWithRow writer factory for deletes with non-null 'row'
+     * @param writerFactoryWithoutRow writer factory for deletes with null 'row'
+     * @param deleteFileFactory delete file factory
+     * @param targetFileSize target file size
+     * @param dsSchema schema of incoming dataset
+     */
+    DeleteWriter(
+        Table table,
+        SparkFileWriterFactory writerFactoryWithRow,
+        SparkFileWriterFactory writerFactoryWithoutRow,
+        OutputFileFactory deleteFileFactory,
+        long targetFileSize,
+        StructType dsSchema) {
+      this.writerWithRow =
+          new ClusteredPositionDeleteWriter<>(
+              writerFactoryWithRow, deleteFileFactory, table.io(), targetFileSize);
+      this.writerWithoutRow =
+          new ClusteredPositionDeleteWriter<>(
+              writerFactoryWithoutRow, deleteFileFactory, table.io(), targetFileSize);
+      this.positionDelete = PositionDelete.create();
+      this.io = table.io();
+      this.specs = table.specs();
+
+      Types.StructType partitionType = Partitioning.partitionType(table);
+
+      this.specIdOrdinal = dsSchema.fieldIndex(PositionDeletesTable.SPEC_ID);
```

Comment:
We know we will write only one spec/partition at a time. What about propagating spec and partition via the rewrite coordinator? Then we can remove all projection code from here and avoid useless work for every row.

Reply:
I ended up propagating via the ScanTaskSetManager. 

I could go either way, but doing this way does seem a little bit hacky, versus before the code was more useful in more cases.